doc/developer/design/20320328_persist_columnar.md

```diff
@@ -0,0 +1,917 @@
+<td> Date </td>
+<td>
+
+```rust
+struct Date {
+  days: i32,
+}
+```
+
+</td>
+<td>
+
+`PrimitiveArray<i32>`
+
+</td>
+<td>
+
+Directly encode the number of days.
+
+</td>
+
+<tr>
+<td> Time </td>
+<td>
+
+```rust
+struct NaiveTime {
+  secs: u32,
+  frac: u32,
+}
+```
+
+</td>
+<td>
+
+`PrimitiveArray<u64>`
+
+</td>
+<td>
+
+Represented as the number of nanoseconds since midnight.
+
+> **Alternative:** Instead of representing this as the number of nanoseconds
+since midnight we could stich together the two `u32`s from `NaiveTime`. This
+would maybe save some CPU cycles during encoding and decoding, but probably not
+enough to matter, and it locks us into a somewhat less flexible format.
+
+> Note: We only need 47 bits to represent this total range, leaving 19 bits
+unused. In the future if we support the `TIMETZ` type we could probably also
+represent that in a `u64`, using these extra bits to store the timezone.
+
+</td>
+
+<tr>
+<td> Timestamp </td>
+<td>
+
+```rust
+struct NaiveDateTime {
+  date: NaiveDate {
+    // (year << 13) | day
+    ymdf: i32,
+  },
+  time: NaiveTime,
+}
+```
+
+</td>
+<td>
+
+`PrimitiveArray<i64>`
+
+</td>
+<td>
+
+`chrono` (our underlying date time library) uses a more memory efficient
+encoding of date by squeezing both year and day into a single `i32`, combined
+with a `NaiveTime` this ends up being 12 bytes.
+
+
+We can repesent this same range of time as the number of microseconds since
+the UNIX epoch in an `i64`. Postgres does something very similar, the only
+difference is it uses an offset of 2000-01-01.
+
+</td>
+
+<tr>
+<td> TimestampTz </td>
+<td>
+
+```rust
+struct DateTime<Tz: TimeZone> {
+    datetime: NaiveDateTime,
+    // purely type info
+    offset: Tz::Offset,
+}
+```
+
+</td>
+<td>
+
+`PrimitiveArray<i64>`
+
+</td>
+<td>
+
+Just like Timestamp, we'll encode this as the number of microseconds since
+the UNIX epoch. We don't actually need to store any timezone information,
+instead we convert to the session timezone when loaded. This is how both
+Postgres works.
+
+</td>
+
+<tr>
+<td> Interval </td>
+<td>
+
+```rust
+struct Interval {
+  months: i32,
+  days: i32,
+  micros: i64,
+}
+```
+
+</td>
+<td>
+
+`FixedSizeBinary[16]`
+
+</td>
+<td>
+
+Represented by encoding the `months`, `days`, and `micros` fields encoded as
+big endian.
+
+> **Alternative:** The smallest possible representation for interval would be
+11 bytes, or 12 if we don't want to do bit swizzling. But other than space
+savings I don't believe there is a benefit to this approach. In fact it would
+incur some computational overhead to encode and there are no benefits from a
+SIMD perspective either.
+
+> **Alternative:** We could represent `Interval`s in a `StructArray` but we
+don't expose the internal details of `Interval` so this wouldn't aid in
+filtering or pushdown. The only benefit of structuring an interval would be for
+space reduction in dictionary encoding, but this is offset by the overhead of
+the struct itself.
```

Comment:
What's the overhead of the struct? I would have expected structs to be ~free because the types are fixed.

Reply:
Arrow Struct Arrays have a validity bitmap for each of their child arrays, which is the overhead I was thinking of. But it turns out that `arrow2` and `arrow-rs` elide these arrays if all of the values are non-null so there isn't actually any overhead.